src/flow_graph.py

```diff
@@ -1056,95 +1159,129 @@ def duplicate_premature_returns(nodes: List[Node]) -> List[Node]:
+    https://www.cs.princeton.edu/courses/archive/spr04/cos598C/lectures/02-ControlFlow.pdf
+    """
+
+    def compute_dominators(
+        entry: Node,
+        parents: Callable[[Node], List[Node]],
+        dominators: Callable[[Node], Set[Node]],
+        immediately_dominates: Callable[[Node], List[Node]],
+        set_immediate_dominator: Callable[[Node, Optional[Node]], None],
+    ) -> None:
+        # See https://en.wikipedia.org/wiki/Dominator_(graph_theory)#Algorithms
+        # Note: if `n` is unreachable from `entry`, then *every* node will
+        # vacuously belong to `n`'s dominator set.
+        for n in nodes:
+            dominators(n).clear()
+            if n == entry:
+                dominators(n).add(n)
             else:
-                fallthrough = None
-            if isinstance(node, BasicNode):
-                if node.emit_goto and fallthrough is not None:
-                    reachable.add(fallthrough)
-                if not node.emit_goto and not node.is_loop():
-                    reachable.add(node.successor)
-            elif isinstance(node, ConditionalNode):
-                assert fallthrough is not None
-                assert fallthrough == node.fallthrough_edge
-                reachable.add(fallthrough)
-                if not node.emit_goto and not node.is_loop():
-                    reachable.add(node.conditional_edge)
-            elif isinstance(node, SwitchNode):
-                assert fallthrough is not None
-                reachable.add(fallthrough)
-            else:  # ReturnNode
-                if node.emit_goto and fallthrough is not None:
-                    reachable.add(fallthrough)
-
-        # We can make the first node not in the set be included in the set by
-        # making its predecessor fall through, and then repeat. For efficiency
-        # we actually do this for all nodes not in the set, but note that we
-        # still need the repeat, since marking nodes fallthrough can cause
-        # edges to disappear.
-        #
-        # At each point we mark more and more nodes fallthrough, so eventually
-        # this process will terminate.
-        assert reachable != last_reachable, "Fallthrough process hit a cycle"
-        last_reachable = reachable
-        unreachable = set(nodes).difference(reachable)
-        if not unreachable:
-            break
-        for node in unreachable:
-            assert node in predecessors, "Start node is never unreachable"
-            pre = predecessors[node]
-            assert not isinstance(pre, ConditionalNode)
-            pre.emit_goto = True
-
-
-def compute_dominators(nodes: List[Node]) -> None:
+                dominators(n).update(nodes)
+
+        changes = True
+        while changes:
+            changes = False
+            for n in nodes:
+                if n == entry:
+                    continue
+                nset = dominators(n)
+                for p in parents(n):
```

Comment:
nit (just noticed this was in the original code): this would be difficult to debug in pdb due to the use of `n` and `p`. Maybe add a TODO to change these variable names to `node` and `parent` respectively?

Reply:
nbd to just fix it now - I'l rename them to `node` & `parent` (and `c` -> `child` for consistency)